### PR TITLE
Token Docs

### PR DESCRIPTION
**Pull request by MartinMinkov**

### CODE REVIEW

**Review by jasongitmail — CHANGES_REQUESTED** [review]
Thanks for this! Left some comments, but requested changes b/c we need to move this to teh Advanced Concepts page once that's merged with the events section.

File: pages/en/zkapps/how-to-write-a-zkapp.mdx [hunk 1]
```diff
@@ -396,6 +394,99 @@ Off-chain state is not yet available.
 
 </Alert>
 
+### Custom Tokens
+
+Custom tokens are a way to create new types of tokens that can be used in the zkApp ecosystem. Custom tokens are identified with a unique "token id" assigned to every kind of token when they are newly minted.
```
[the remark this answers is not in the document]
Comment (iregina) [reply pages/en/zkapps/how-to-write-a-zkapp.mdx:399]: +1

File: pages/en/zkapps/how-to-write-a-zkapp.mdx [hunk 2]
```diff
@@ -396,6 +394,99 @@ Off-chain state is not yet available.
...
+
+Custom tokens are typically governed by zkApp accounts, where zkApp accounts can mint/burn tokens and facilitate transfers of two separate accounts. zkApps developers can use the supported SnarkyJS API for custom tokens to send, burn, and authorize transactions between other accounts.
+
```
Comment (iregina) [inline pages/en/zkapps/how-to-write-a-zkapp.mdx:402]: Similarly to the phrasing above, one suggestion: 

"You can interact with custom tokens using the SnarkyJS API. SnarkyJS contains built-in methods to send, burn, and authorize transactions between accounts"

File: pages/en/zkapps/how-to-write-a-zkapp.mdx [hunk 3]
```diff
@@ -396,6 +394,99 @@ Off-chain state is not yet available.
...
+##### Custom Token API
+
+There are three main methods for zkApps to interact with custom tokens. This section will discuss how minting, burning, and sending custom tokens are exposed to zkApp developers.
+
```
Comment (iregina) [inline pages/en/zkapps/how-to-write-a-zkapp.mdx:406]: Suggestion: "This section shows you how to perform common token operations, such as minting, burning, and sending tokens, using SnarkyJS API."

File: pages/en/zkapps/how-to-write-a-zkapp.mdx [hunk 4]
```diff
@@ -396,6 +394,99 @@ Off-chain state is not yet available.
...
+##### Burning
+
+Burning tokens is the opposite of minting. Burning tokens deductes the balance of a certain address by the specified amount. See the example below of how a zkApp can burn tokens of another account:
+
+```ts
+class BurnExample extends SmartContract {
+...
+@method burnTokens(receiverAddress: PublicKey) {
+    this.token().burn({
+      address: receiverAddress,
```
[the remark this answers is not in the document]
Comment (iregina) [reply pages/en/zkapps/how-to-write-a-zkapp.mdx:436]: More for my understanding, if you burn a token, does it just disappear? OR does it get moved to another address? Can you set permissions to who is allowed to burn or create tokens?

File: pages/en/zkapps/how-to-write-a-zkapp.mdx [hunk 5]
```diff
@@ -396,6 +394,99 @@ Off-chain state is not yet available.
...
+
+If you are writing your own zkApp to interact with custom tokens, you should understand the following important terms:
+
+##### Token Accounts
+
+Token accounts are like regular accounts, but they hold a balance of a specific custom token instead of MINA. They are created from existing accounts and are specified by a public key and token id.
+If an existing account receives a transaction that is specified by a custom token, a token account for that public key and token id will be created if it does not exist.
+Token accounts are specific for each type of custom token, meaning that a single public key can have many different types of token accounts.
+
+To create a token account for a specified custom token, whenever an existing account receives a transaction denoted with a custom token, a token account is automatically created for that public key.
```
Comment (iregina) [inline pages/en/zkapps/how-to-write-a-zkapp.mdx:482]: Does the account creation fee apply here? If so, maybe we should add a note?

File: pages/en/zkapps/how-to-write-a-zkapp.mdx [hunk 6]
```diff
@@ -396,6 +394,99 @@ Off-chain state is not yet available.
...
+
+Aside from sending custom tokens, custom tokens can be minted and burned by a **token owner account**. A token owner account is the governing zkApp account for a specific custom token.
```
Comment (iregina) [inline pages/en/zkapps/how-to-write-a-zkapp.mdx:484]: So there are two types of accounts? Shouldn't we call one token holder account and the other token owner account? (sorry if you already discussed this -- this is all brand new to me)

File: pages/en/zkapps/how-to-write-a-zkapp.mdx [hunk 7]
```diff
@@ -396,6 +394,99 @@ Off-chain state is not yet available.
...
+
+##### Token Owner
+
+A token owner is an account that facilitates and governs how a custom token is to be used. Custom tokens are not created by sending a transaction to the network, but rather they are _derived_ from an existing account.
+
 ## Next Steps
 
-Now that you've learn how to write a smart contract, you can now learn <DocLink copy="how to test your zkApp" url="/zkapps/how-to-test-a-zkapp" />.
\ No newline at end of file
+Now that you've learn how to write a smart contract, you can now learn <DocLink copy="how to test your zkApp" url="/zkapps/how-to-test-a-zkapp" />.
```
Comment (iregina) [inline pages/en/zkapps/how-to-write-a-zkapp.mdx:492]: Excellent use of spacing here! I can hear your voice when reading this. GREAT JOB!!!!!!!!!!!!!!! (just wanted to leave a positive non-critical comment)

File: pages/en/zkapps/how-to-write-a-zkapp.mdx (hunk 4 above)
[the remark this answers is not in the document]
Comment (jasongitmail) [reply pages/en/zkapps/how-to-write-a-zkapp.mdx:436]: @iregina the token balance is reduced on the specified address--it just disappears. Only the Token Owner can burn or mint.

File: pages/en/zkapps/how-to-write-a-zkapp.mdx (hunk 1 above)
[the remark this answers is not in the document]
Comment (MartinMinkov) [reply pages/en/zkapps/how-to-write-a-zkapp.mdx]: Fixed!

File: pages/en/zkapps/how-to-write-a-zkapp.mdx [hunk 8]
```diff
@@ -396,6 +394,99 @@ Off-chain state is not yet available.
...
+
+To mint new tokens using a zkApp, we access the `this.token` property on our SmartContract class. See the example below of how a zkApp can mint tokens to another account:
+
+```ts
+class MintExample extends SmartContract {
+...
+@method mintNewTokens(receiverAddress: PublicKey) {
+    this.token().mint({
+      address: receiverAddress,
+      amount: 1000000,
```
[the remark this answers is not in the document]
Comment (MartinMinkov) [reply pages/en/zkapps/how-to-write-a-zkapp.mdx]: Fixed the issue with the number formatting.

As for the `amount` vs `quantity`, the `amount` property is specified by the SnarkyJS API so we can't really change that name here. If we want to rename this property in SnarkyJS, we should probably bring this up in a discussion and let other team members know.

File: pages/en/zkapps/how-to-write-a-zkapp.mdx [hunk 9]
```diff
@@ -396,6 +394,99 @@ Off-chain state is not yet available.
...
+    });
+  }
+}
+```
+
+In the snippet above, we define a smart contract called `BurnExample` with a method called `burnTokens`. Similar to minting, we use the `this.token()` property to call the `burn` method. This specifies the amount of tokens to burn for the specified address.
+
+**Note**: A zkApp cannot burn more tokens than the specified account has. If there is such a case, an error is thrown and no such transaction is made.
```
[the remark this answers is not in the document]
Comment (MartinMinkov) [reply pages/en/zkapps/how-to-write-a-zkapp.mdx]: Added!

File: pages/en/zkapps/how-to-write-a-zkapp.mdx [hunk 10]
```diff
@@ -396,6 +394,99 @@ Off-chain state is not yet available.
...
+
+##### Sending
+
+If two accounts hold a balance in a custom token, a zkApp must authorize such transaction. Sending tokens requires authorization from the zkApp, which is done by calling the `send` method on the `token` property. See the example below of how a zkApp can authorize sending tokens between two accounts:
```
[the remark this answers is not in the document]
Comment (MartinMinkov) [reply pages/en/zkapps/how-to-write-a-zkapp.mdx]: Fixed!

File: pages/en/zkapps/how-to-write-a-zkapp.mdx [hunk 11]
```diff
@@ -396,6 +394,99 @@ Off-chain state is not yet available.
...
+      from: senderAddress,
+      amount,
+    });
+  }
+}
+```
+
+In the snippet above, we define a smart contract called `SendExample` with a method called `sendTokens`. In the same fashion as minting and burning, we use the `this.token()` property to call the `send` method.
+
+By defining a send method on our smart contract, the zkApp will fill in the authorization it needs to faciltate the transfer between two seperate accounts. Once that has been supplied, the two accounts simply add/subtract the specified amount from their respective balances.
```
[the remark this answers is not in the document]
Comment (MartinMinkov) [reply pages/en/zkapps/how-to-write-a-zkapp.mdx]: I got rid of this paragraph here. I was trying to mention the fact that token transfers need to be authorized by a token owner but this came off bad. I removed it and mentioned it more in a later section.

File: pages/en/zkapps/how-to-write-a-zkapp.mdx (hunk 2 above)
Comment (MartinMinkov) [reply pages/en/zkapps/how-to-write-a-zkapp.mdx]: Fixed!

File: pages/en/zkapps/how-to-write-a-zkapp.mdx (hunk 5 above)
Comment (MartinMinkov) [reply pages/en/zkapps/how-to-write-a-zkapp.mdx]: That's a great point! Added!

File: pages/en/zkapps/how-to-write-a-zkapp.mdx (hunk 6 above)
Comment (MartinMinkov) [reply pages/en/zkapps/how-to-write-a-zkapp.mdx]: Yup, that's right! I am trying to stick with the terminology we use in other parts of the protocol so I'm a little hesitant to introduce a new term like `token holder account`. I rephrased the content a bit to try and get the point across better.

**Comment by jasongitmail** [comment]
Closing because this content was integrated onto the new docs website already https://github.com/o1-labs/docs2

**Pull request closed without merging**